src/snowflake/connector/query_context_cache.py

```diff
@@ -0,0 +1,302 @@
+        if not isinstance(other, QueryContextElement):
+            return False
+        return (
+            self._id == other.id
+            and self._read_timestamp == other.read_timestamp
+            and self._priority == other.priority
+            and self._context == other.context
+        )
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            raise NotImplementedError(
+                f"cannot compare QueryContextElement with object of type {type(other)}"
+            )
+        return self._priority < other._priority
+
+    def __hash__(self) -> int:
+        _hash = 31
+
+        _hash = _hash * 31 + self._id
+        _hash += (_hash * 31) + self._read_timestamp
+        _hash += (_hash * 31) + self._priority
+        if self._context:
+            _hash += (_hash * 31) + int.from_bytes(
+                md5(self._context.encode("utf-8")).digest(), "big"
+            )
+        return _hash
+
+    def __str__(self) -> str:
+        return f"({self._id}, {self._read_timestamp}, {self._priority})"
+
+
+class QueryContextCache:
+    def __init__(self, capacity: int):
+        self._capacity = capacity
+        self._id_map: dict[int, QueryContextElement] = {}
+        self._priority_map: dict[int, QueryContextElement] = {}
+        self._intermediate_priority_map: dict[int, QueryContextElement] = {}
+
+        # stores elements sorted by priority. Element with
+        # least priority value has the highest priority
+        self._tree_set: set[QueryContextElement] = SortedSet()
+        self._lock = Lock()
+        self._data: str = None
+
+    @property
+    def capacity(self) -> int:
+        return self._capacity
+
+    def _add_qce(self, qce: QueryContextElement) -> None:
+        self._tree_set.add(qce)
+        self._id_map[qce.id] = copy.deepcopy(qce)
+        self._intermediate_priority_map[qce.priority] = copy.deepcopy(qce)
+
+    def _remove_qce(self, qce: QueryContextElement) -> None:
+        self._id_map.pop(qce.id)
+        self._priority_map.pop(qce.priority)
+        self._tree_set.remove(qce)
+
+    def _replace_qce(
+        self, old_qce: QueryContextElement, new_qce: QueryContextElement
+    ) -> None:
+        self._remove_qce(old_qce)
+        self._add_qce(new_qce)
+
+    def _sync_priority_map(self):
+        """
+        Sync the _intermediate_priority_map with the _priority_map at the end of the current round of merges.
+        """
+        logger.debug(
+            f"syncPriorityMap called priority_map size = {len(self._priority_map)}, new_priority_map size = {len(self._intermediate_priority_map)}"
+        )
+
+        self._priority_map.update(self._intermediate_priority_map)
+        # Clear the _intermediate_priority_map for the next round of QCC merge (a round consists of multiple entries)
+        self._intermediate_priority_map.clear()
+
+    def merge(self, id: int, read_timestamp: int, priority: int, context: str) -> None:
```

Comment:
I could call it `raw_merge`. We didn't because merge is only called with individual values and since not we don't merge an element in each case, creating a new QCC element is slightly wasteful

Reply:
calling it `insert` based on your suggestion